### PR TITLE
Fix/csi 1540 already exists gap create volume

### PR DESCRIPTION
**Pull request by ArbelNathan**

### CODE REVIEW

**Review by oriyarde — APPROVED** [review]
LGTM

**Comment by oriyarde** [comment]
@ArbelNathan btw could you add a UT to cover it (in the same/a separate ticket)?

**Review by lisyukanna — DISMISSED** [review]
LGTM

**Review by oriyarde — DISMISSED** [review]
nice job!

File: controller/tests/controller_server/csi_controller_server_test.py [hunk 1]
```diff
@@ -532,6 +545,62 @@ def test_create_volume_with_zero_size(self, storage_agent):
...
+        self.mediator.get_volume.return_value = utils.get_mock_mediator_response_volume(10, volume_name, "wwn2",
+                                                                                        "a9k")
+        response = self.servicer.CreateVolume(self.request, self.context)
+
+        self.assertEqual(self.context.code, grpc.StatusCode.ALREADY_EXISTS)
+        self.assertEqual(response.volume.content_source.snapshot.snapshot_id, '')
+        self.mediator.copy_to_existing_volume_from_source.assert_not_called()
+
+    @patch("controller.controller_server.csi_controller_server.get_agent")
+    def test_create_volume_from_source_idempotent_source_not_requested_but_found(self, storage_agent):
```
Comment (oriyarde) [inline controller/tests/controller_server/csi_controller_server_test.py:580]: then why `from_source`?

File: controller/tests/controller_server/csi_controller_server_test.py [hunk 2]
```diff
@@ -532,6 +545,62 @@ def test_create_volume_with_zero_size(self, storage_agent):
...
+        self.mediator.get_volume.return_value = utils.get_mock_mediator_response_volume(10, volume_name, "wwn2", "a9k",
+                                                                                        copy_source_id=snapshot_id)
+        response = self.servicer.CreateVolume(self.request, self.context)
+
+        self.assertEqual(self.context.code, grpc.StatusCode.ALREADY_EXISTS)
+        self.assertEqual(response.volume.content_source.snapshot.snapshot_id, '')
+        self.mediator.copy_to_existing_volume_from_source.assert_not_called()
+
+    @patch("controller.controller_server.csi_controller_server.get_agent")
+    def test_create_volume_from_source_error_other_source(self, storage_agent, ):
```
Comment (oriyarde) [inline controller/tests/controller_server/csi_controller_server_test.py:594]: why the extra comma?

File: controller/tests/controller_server/csi_controller_server_test.py [hunk 3]
```diff
@@ -532,6 +545,62 @@ def test_create_volume_with_zero_size(self, storage_agent):
...
+        self._prepare_idempotent_tests()
+        storage_agent.return_value = self.storage_agent
+        snapshot_id = "wwn1"
+        self.request.volume_content_source = None
+        self.mediator.get_volume.return_value = utils.get_mock_mediator_response_volume(10, volume_name, "wwn2", "a9k",
+                                                                                        copy_source_id=snapshot_id)
+        response = self.servicer.CreateVolume(self.request, self.context)
+
+        self.assertEqual(self.context.code, grpc.StatusCode.ALREADY_EXISTS)
+        self.assertEqual(response.volume.content_source.snapshot.snapshot_id, '')
```
Comment (oriyarde) [inline controller/tests/controller_server/csi_controller_server_test.py:590]: shouldn't there be an id value in this case?
and if not, should the response even contain a volume?
and if not, then why is the assert so deep in the response object?

Comment (ArbelNathan) [reply controller/tests/controller_server/csi_controller_server_test.py:590]: [spec ](https://github.com/container-storage-interface/spec/blob/v1.3.0/spec.md#createvolume)specify new created so "old" volume wont return

File: controller/tests/controller_server/csi_controller_server_test.py (hunk 1 above)
Comment (ArbelNathan) [reply controller/tests/controller_server/csi_controller_server_test.py]: done

File: controller/tests/controller_server/csi_controller_server_test.py (hunk 2 above)
Comment (ArbelNathan) [reply controller/tests/controller_server/csi_controller_server_test.py]: done

File: controller/tests/controller_server/csi_controller_server_test.py (hunk 3 above)
Comment (ArbelNathan) [reply controller/tests/controller_server/csi_controller_server_test.py]: done (changed to `HasField`)

File: controller/tests/controller_server/csi_controller_server_test.py (hunk 2 above)
Comment (oriyarde) [reply controller/tests/controller_server/csi_controller_server_test.py]: 1. `test_create_idempotent` should be `test_create_volume_idempotent` here.
2. now that `test_create_volume_from_source_idempotent_no_source` was changed to `test_create_volume_idempotent_with_source_volume_have_no_source`, IMO it is more ambiguous (request vs. existing volume) - perhaps `idempotent` is not a must, maybe just `test_create_volume_from_source_already_exists_without_source`?

File: controller/tests/controller_server/csi_controller_server_test.py (hunk 1 above)
Comment (oriyarde) [reply controller/tests/controller_server/csi_controller_server_test.py]: 1. `test_create_idempotent` should be `test_create_volume_idempotent` here.
2. now that `test_create_volume_from_source_idempotent_no_source` was changed to `test_create_volume_idempotent_with_source_volume_have_no_source`, IMO it is more ambiguous (request vs. existing volume) - perhaps `idempotent` is not a must, maybe just `test_create_volume_from_source_already_exists_without_source`?

File: controller/tests/controller_server/csi_controller_server_test.py (hunk 3 above)
Comment (oriyarde) [reply controller/tests/controller_server/csi_controller_server_test.py]: so why not [here](https://github.com/IBM/ibm-block-csi-driver/blob/0ccea31ebcc4c1b34b97674c0d065b8027ab2a17/controller/tests/controller_server/csi_controller_server_test.py#L576) also?

File: controller/tests/controller_server/csi_controller_server_test.py (hunk 1 above)
Comment (ArbelNathan) [reply controller/tests/controller_server/csi_controller_server_test.py]: but its a specific test for idempotent

Comment (ArbelNathan) [reply controller/tests/controller_server/csi_controller_server_test.py]: 1 is done

File: controller/tests/controller_server/csi_controller_server_test.py (hunk 3 above)
Comment (ArbelNathan) [reply controller/tests/controller_server/csi_controller_server_test.py]: right done

**Review by oriyarde — APPROVED** [review]
great job once again!

File: controller/tests/controller_server/csi_controller_server_test.py (hunk 3 above)
Comment (oriyarde) [reply controller/tests/controller_server/csi_controller_server_test.py]: and [here](https://github.com/IBM/ibm-block-csi-driver/blob/0b40468bbb83f62464aa6a483ad080ad590d91a7/controller/tests/controller_server/csi_controller_server_test.py#L403)?

Comment (ArbelNathan) [reply controller/tests/controller_server/csi_controller_server_test.py]: yep

Comment (oriyarde) [reply controller/tests/controller_server/csi_controller_server_test.py]: I don't see the change in my last link

**Review by lisyukanna — APPROVED** [review]
LGTM

File: controller/controller_server/csi_controller_server.py [hunk 4]
```diff
@@ -189,7 +189,7 @@ def _handle_existing_volume_source(self, volume, source_id, source_type, context
         """
         Args:
             volume              : volume fetched or created in CreateVolume
-            source_id       : id of object we should copy to vol or None if volume should not be copied
+            source_id           : id of object we should copy to vol or None if volume should not be copied
             source_type:        : the object type of the source - volume or snapshot
             context             : CreateVolume response context
         Returns:
```
Comment (oriyarde) [inline controller/controller_server/csi_controller_server.py:195]: > If volume is a copy of another source - set context status to INTERNAL...

consider removing this docstring altogether

**Pull request merged**